lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +796,585 @@ def __str__(self):
+
+
+class GitTag(GitRef):
+    """A git tag."""
+
+    ref_type = 'Tag'        # type: ClassVar[str]
+
+    @property
+    def version(self):
+        # type: () -> spack.version.Version
+        return spack.version.Version(self._ref)
+
+    @property
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        return self._commit
+
+    def __init__(self, ref, commit=None):
+        # type: (str, Optional[str]) -> None
+        super(GitTag, self).__init__(ref)
+        self._commit = commit
+
+    def __repr__(self):
+        return 'GitTag(ref_type={!r}, ref={!r}, commit={!r})'.format(
+            self.ref_type, self._ref, self._commit,
+        )
+
+GitRef.Tag = GitTag             # noqa: E305
+
+
+class GitBranch(GitRef):
+    """A git branch."""
+
+    ref_type = 'Branch'        # type: ClassVar[str]
+
+    @property
+    def should_validate_matches_hash(self):
+        # type: () -> Optional[str]
+        return self._commit
+
+    def __init__(self, ref, commit=None):
+        # type: (str, Optional[str]) -> None
+        super(GitBranch, self).__init__(ref)
+        self._commit = commit
+
+    def __repr__(self):
+        return 'GitBranch(ref_type={!r}, ref={!r}, commit={!r})'.format(
+            self.ref_type, self._ref, self._commit,
+        )
+
+GitRef.Branch = GitBranch       # noqa: E305
+
+
+class GitFetchStageConfiguration(object):
+    """Validate parameters used to customize a specific git fetch operation.
+
+    This information is orthogonal to that which is processed in `GitRef`. It describes
+    what operations are performed to *prepare* a git checkout for a spack `Stage`,
+    *after* fetching the configured ref.
+    """
+    submodules = None           # type: bool
+    submodules_delete = None    # type: Optional[List[str]]
+    get_full_repo = None        # type: bool
+
+    def __repr__(self):
+        return '{0}(submodules={1}, submodules_delete={2}, get_full_repo={3})'.format(
+            type(self).__name__,
+            self.submodules,
+            self.submodules_delete,
+            self.get_full_repo,
+        )
+
+    def __init__(self, submodules, submodules_delete, get_full_repo):
+        # type: (bool, Optional[List[str]], bool) -> None
+        self.submodules = submodules
+        self.submodules_delete = submodules_delete
+        self.get_full_repo = get_full_repo
+
+    @classmethod
+    def _extract_bool(cls, kwargs, name, default):
+        # type: (Any, str, bool) -> bool
+        value = kwargs.pop(name, default)
+        if isinstance(value, bool):
+            return value
+        raise InvalidGitFetchStageConfig(
+            'argument {0}={1!r} must be a bool'.format(name, value))
+
+    @classmethod
+    def from_version_directive(cls, kwargs):
+        # type: (Any) -> GitFetchStageConfiguration
+        try:
+            submodules = cls._extract_bool(kwargs, 'submodules', False)
+            submodules_delete = kwargs.pop('submodules_delete', None)
```

Comment:
This should default to None no?

Reply:
Yes, and I believe that's what `kwargs.pop('submodules_delete', None)` is doing?